dulwich/tests/test_porcelain.py

```diff
@@ -630,9 +630,12 @@ def test_no_head_no_checkout_outstream_errstream_autofallback(self):
         r.close()
 
     def test_source_broken(self):
-        target_path = tempfile.mkdtemp()
-        self.assertRaises(Exception, porcelain.clone, "/nonexistant/repo", target_path)
-        self.assertFalse(os.path.exists(target_path))
+        with tempfile.TemporaryDirectory() as parent:
+            target_path = os.path.join(parent, "target")
```

Comment:
With the adjusted `rmtree` behavior, the target repo directory will only be removed when `mkdir=True` (which for porcelain is when the directory does not exist before `porcelain.clone()` is called). So `target_path` can't be the dir generated by `mkdtemp()` `target_path` has to be a child of that dir instead.

I used `TemporaryDirectory` instead of `mkdtemp` since the context manager will clean everything up afterwards whether or not the test succeeds.